pkg/k8s/factory.go

```diff
@@ -0,0 +1,245 @@
+//    or you stop the controller).
+//  * rehf: is the  resource event handler funcs that is used to handle the
+//    stream events.
+func ControllerFactory(
+	k8sGetter cache.Getter,
+	resourceObj runtime.Object,
+	resyncPeriod time.Duration,
+	rehf cache.ResourceEventHandlerFuncs,
+) cache.Controller {
+
+	_, controller := cache.NewInformer(
+		cache.NewListWatchFromClient(
+			k8sGetter,
+			resourceNameOf(resourceObj),
+			v1.NamespaceAll,
+			fields.Everything(),
+		),
+		resourceObj,
+		resyncPeriod,
+		rehf,
+	)
+
+	return controller
+}
+
+// ResourceEventHandlerFactory returns a kubernetes ResourceEventHandlerFuncs,
+// the resource event handler will have a serializer.FunctionQueue to enqueue
+// the received events.
+// Parameters:
+//  * addFunc is the function serialized in the queue when an object addition is
+//    received.
+//  * delFunc is the function serialized in the queue when an object deletion is
+//    received.
+//  * updateFunc is the function serialized in the queue when an object update
+//    is received.
+//  * resourceObj: is an object of the type that you expect to receive.
+func ResourceEventHandlerFactory(
+	addFunc, delFunc func(i interface{}) func() error,
+	updateFunc func(old, new interface{}) func() error,
+	resourceObj runtime.Object,
+) cache.ResourceEventHandlerFuncs {
+
+	fqueue := serializer.NewFunctionQueue(1024)
+	castToDeepCopy := castFuncFactory(resourceObj)
+
+	rehf := cache.ResourceEventHandlerFuncs{}
+	if addFunc != nil {
+		rehf.AddFunc = func(obj interface{}) {
+			metrics.SetTSValue(metrics.EventTSK8s, time.Now())
+			if metaObj := castToDeepCopy(obj); metaObj != nil {
+				fqueue.Enqueue(addFunc(metaObj), serializer.NoRetry)
+			}
+		}
+	}
+	if updateFunc != nil {
+		rehf.UpdateFunc = func(oldObj, newObj interface{}) {
+			metrics.SetTSValue(metrics.EventTSK8s, time.Now())
+			if oldMetaObj := castToDeepCopy(oldObj); oldMetaObj != nil {
+				if newMetaObj := castToDeepCopy(newObj); newMetaObj != nil {
+					fqueue.Enqueue(updateFunc(oldMetaObj, newMetaObj), serializer.NoRetry)
+				}
+			}
+		}
+	}
+
+	if delFunc != nil {
+		rehf.DeleteFunc = func(obj interface{}) {
+			metrics.SetTSValue(metrics.EventTSK8s, time.Now())
+			if metaObj := castToDeepCopy(obj); metaObj != nil {
+				fqueue.Enqueue(delFunc(metaObj), serializer.NoRetry)
+			}
+		}
+	}
+
+	return rehf
+}
+
+func resourceNameOf(i interface{}) string {
+	switch i.(type) {
+	case *networkingv1.NetworkPolicy:
+		return "networkpolicies"
+	case *v1.Service:
+		return "services"
+	case *v1.Endpoints:
+		return "endpoints"
+	case *v1beta1.Ingress:
+		return "ingresses"
+	case *cilium_v2.CiliumNetworkPolicy:
+		return "ciliumnetworkpolicies"
+	case *v1.Pod:
+		return "pods"
+	case *v1.Node:
+		return "nodes"
+	case *v1.Namespace:
+		return "namespaces"
+	default:
+		log.Panicf("Invalid resource type %s", reflect.TypeOf(i))
+		return ""
+	}
+}
+
+func castFuncFactory(i interface{}) func(i interface{}) meta_v1.Object {
```

Comment:
This function should be called `castToDeepCopy`. We only use it in one place anyway (where it's called that) and that's what it does (more or less...)